check for invalid ranges in torch.arange

nairbv opened the pull request:
#13782

nairbv committed:
check for invalid ranges in torch.arange

soumith commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -77,6 +77,10 @@ Tensor arange(
     Scalar step,
     const TensorOptions& options) {
+  AT_CHECK(
+    finite(end.toDouble()) && finite(start.toDouble()),
```
I think you want to use `isfinite` in C99/C++11. `finite` is Linux-specific

nairbv replied:
ah, updated. would our tests catch this?

soumith replied:
yea, the CI caught it, Windows build failed :D

zou3519 commented on test/test_torch.py:
```diff
@@ -2981,6 +2981,13 @@ def test_arange(self):
 
+        self.assertRaises(RuntimeError, lambda: torch.arange(0, float('nan')))
+        self.assertRaises(RuntimeError, lambda: torch.arange(0, float('-inf')))
+        self.assertRaises(RuntimeError, lambda: torch.arange(0, float('inf')))
```
nit: self.assertRaisesRegex (with a regex on unsupported range) would probably be better. I can imagine a (wrong) scenario where we attempt to allocate memory for float('inf') and instead report OOM instead of "unsupported range".

nairbv replied:
fixed

zou3519 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -77,6 +77,10 @@ Tensor arange(
     Scalar step,
     const TensorOptions& options) {
+  AT_CHECK(
+    finite(end.toDouble()) && finite(start.toDouble()),
```
nit: std::is_finite might be more portable, but I don't know

nairbv replied:
updated, @soumith mentioned this above as well.

nairbv committed:
torch.arange range check updates from review #13782

vadimkantorov commented:
Does the fix affect the overload `torch.arange(end)`? The corresponding ATen code seems to call in the legacy function without the check: https://github.com/pytorch/pytorch/blob/f472a5075c83b26a5f1177b214ce3f0d691197c3/aten/src/ATen/native/TensorFactories.cpp#L98

The same for all `arange_out` overloads

nairbv committed:
move range check to THTensorMoreMath so applicable to all arange calls

vadimkantorov commented:
@nairbv I don't understand PyTorch's internals very well. Is `_th_arange_out` also handled by (the now fixed) `THTensor_(arange)(THTensor *r_, accreal xmin, accreal xmax, accreal step)`?

ezyang commented:
Appears so:

```
Tensor & CPUFloatType::_th_arange_out(Tensor & result, Scalar start, Scalar end, Scalar step) const {
    // DeviceGuard omitted
    auto result_ = checked_tensor_unwrap(result,"result",0, false, Backend::CPU, ScalarType::Float);
    auto start_ = start.toDouble();
    auto end_ = end.toDouble();
    auto step_ = step.toDouble();
    THFloatTensor_arange(result_, start_, end_, step_);
    return result;
}
```

nairbv commented:
> @nairbv I don't understand PyTorch's internals very well. Is `_th_arange_out` also handled by (the now fixed) `THTensor_(arange)(THTensor *r_, accreal xmin, accreal xmax, accreal step)`?

yes, I'm new to this as well but the easiest way to see it is if you build the code, it'll appear in the generated file CPUFloatType.cpp (generated by aten/src/ATen/gen.py).

All of the `_th_arange*` functions call THFLoatTensor_arange(result_, start_, end_, step_); where sometimes start_ defaults to 0 and sometimes step defaults to 1... where sometimes result_ is a parameter and sometimes result is created. macro expansions lead from the THTensor_(arange) changed here to the THFloatTEnsor_arange being called.

vadimkantorov commented:
Is CUDA version also fixed by the edits in `THTensor_(arange)`?

For before it was, so CUDA was affected too:
```
>>> a = torch.arange(0, float('inf'), device = 'cuda')
>>> a.shape
torch.Size([-9223372036854775808])
```

vadimkantorov commented:
If you move down the guard to TH (as it appears now), it probably should also be duplicated in THC: https://github.com/pytorch/pytorch/blob/86829997676a5025b382bd51a538f3a07ecdf5b0/aten/src/THC/generic/THCTensorMath.cu#L461

nairbv committed:
improve parameter checking for torch.arange

added parameter checks to cuda version of tensor, used standard
THArgCheck instead of AT_CHECK, updated tests to verify use on both
device types.

vadimkantorov commented on aten/src/THC/generic/THCTensorMath.cu:
```diff
@@ -461,6 +461,8 @@ void THCTensor_(range)(THCState *state, THCTensor *r_, accreal xmin, accreal xma
+              , 1, "unsupported range: ", xmin, " -> ", xmax);
   THArgCheck(((step > 0) && (xmax >= xmin)) || ((step < 0) && (xmax <= xmin))
               , 2, "upper bound and larger bound inconsistent with step sign");
   ptrdiff_t size = (ptrdiff_t) ceil(ScalarConvert<accreal, double>::to(xmax - xmin) / step);
```
Can this still overflow `ptrdiff_t` if we pass something like torch.arange(1.175494351e-38, 3.402823466e+38)? Can `xmax-xmin` itself turn to `inf` if `xmin` is negative and `xmax - xmin` does not fit in double?

vadimkantorov replied:
Even with good finite non-infs:
```
torch.arange(1.175494351e-38, 3.402823466e+38).shape
torch.Size([-9223372036854775808])
```

nairbv replied:
hmm... it seems like overflow is supposed to be caught by THGeneral.cpp:191:

```
>>> x = torch.arange(0, 9.223372036854775E18)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: $ Torch: invalid memory size -- maybe an overflow? at /home/bvaughan/repos/pytorch/aten/src/TH/THGeneral.cpp:191

```

nairbv committed:
handle overflow of numel to avoid negative size

vadimkantorov commented:
Can an overflow result in a positive value or even a zero? Probably not?

nairbv commented:
> Can an overflow result in a positive value or even a zero? Probably not? (depends on how many times the overflow rolls - probably not more than 1?)

hm... it seems that the behavior we're seeing is in the cast from double to int64_t (or actually, to ptrdiff_t). From what I can tell the behavior for an out of range cast is undefined, but in practice it doesn't seem to be wrapping. You saw this -9223372036854775808 value in tests (the minimum long value, or max long + 1).

We could try a range check on the double first. That's tricky though because ptrdiff_t's max value isn't exactly representable in floating point, and could differ in 32 bit systems. I'm not sure if there is actually a system that would wrap it, but I'm also pretty new to C++.

I'm a bit skeptical of the floating point arithmetic in general of `(max-min)/step`, but I don't see a specific issue with it yet, and we might be getting beyond the scope of the original ticket. Maybe we should get this merged, and file another ticket if we find another specific failure case?

nairbv committed:
fix "ambiguous call to overloaded function" in windows CI build

ezyang commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -637,11 +637,15 @@ void THTensor_(arange)(THTensor *r_, accreal xmin, accreal xmax, accreal step) {
   scalar_t i = 0;
 
   THArgCheck(step > 0 || step < 0, 3, "step must be nonzero");
+  THArgCheck(std::isfinite((double)xmin) && std::isfinite((double)xmax)
```
nit: prefer C++ cast (`static_cast<double>`) over C-style cast (`(double)`)

ezyang replied:
Also, it's definitely a bit tricky that you're converting to double here, seems worth a comment.

nairbv commented:
One of the newly-added tests did end up failing in one environment with:

Nov 19 23:30:46 test_arange (__main__.TestTorch) ... /var/lib/jenkins/workspace/aten/src/TH/generic/THTensorMoreMath.cpp:645:22: runtime error: 3.40282e+38 is outside the range of representable values of type 'long'

so I guess we will need to explicitly check the ranges before casting.

nairbv committed:
explicitly check range of ptrdiff_t in arange

nairbv committed:
__PTRDIFF_MAX__ not available on some systems, using PTRDIFF_MAX instead

nairbv commented:
@pytorchbot retest this please

nairbv commented:
@pytorchbot retest this please

ezyang commented:
> I'm a bit skeptical of the floating point arithmetic in general of (max-min)/step, but I don't see a specific issue with it yet, and we might be getting beyond the scope of the original ticket. Maybe we should get this merged, and file another ticket if we find another specific failure case?

I don't think anyone is actually blocked by the lack of a check here. Let's take the time to do it right.

ezyang commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -637,10 +637,16 @@ void THTensor_(arange)(THTensor *r_, accreal xmin, accreal xmax, accreal step) {
   THArgCheck(step > 0 || step < 0, 3, "step must be nonzero");
+  THArgCheck(std::isfinite(static_cast<double>(xmin)) &&
+              std::isfinite(static_cast<double>(xmax))
+              , 1, "unsupported range: ", xmin, " -> ", xmax);
```
I checked the bug report, and the PR comments, but I don't see why you need to cast the quantity to a double before checking if it's finite. I'm pretty sure `std::isfinite` has overloads for the integral types too. Is there something I'm missing?

nairbv replied:
hmm...I thought I remembered getting an "ambiguous call to overloaded function" when I didn't cast, but, not seeing that now.

nairbv replied:
ah, yeah, looks like that 'ambiguous' error only occurs on windows. I'll need to re-add the cast:

```
15:26:10 C:\Program Files (x86)\Windows Kits\10\include\10.0.16299.0\ucrt\corecrt_math.h(402): error C2668: 'fpclassify': ambiguous call to overloaded function
15:26:10 C:\Program Files (x86)\Windows Kits\10\include\10.0.16299.0\ucrt\corecrt_math.h(299): note: could be 'int fpclassify(long double) throw()'
15:26:10 C:\Program Files (x86)\Windows Kits\10\include\10.0.16299.0\ucrt\corecrt_math.h(294): note: or       'int fpclassify(double) throw()'
15:26:10 C:\Program Files (x86)\Windows Kits\10\include\10.0.16299.0\ucrt\corecrt_math.h(289): note: or       'int fpclassify(float) throw()'
15:26:10 C:\Program Files (x86)\Windows Kits\10\include\10.0.16299.0\ucrt\corecrt_math.h(402): note: while trying to match the argument list '(int64_t)'
15:26:10 c:\jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\aten\src\th\generic/THTensorMoreMath.cpp(642): note: see reference to function template instantiation 'bool isfinite<int64_t>(_Ty) throw()' being compiled
15:26:10         with
15:26:10         [
15:26:10             _Ty=int64_t
15:26:10         ]

```

ezyang commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -637,10 +637,16 @@ void THTensor_(arange)(THTensor *r_, accreal xmin, accreal xmax, accreal step) {
+  double size_d = ceil(static_cast<double>(xmax - xmin) / step);
+  THArgCheck(size_d >= 0 && size_d <= static_cast<double>(PTRDIFF_MAX)
+             , 1, "invalid size, possible overflow?");
+  size = static_cast<ptrdiff_t>(size_d);
```
Yeah, the `static_cast<double>(PTRDIFF_MAX)` gives me hives. I guess it's OK to not fix it properly for now, but if you don't get a comment here, all of the discussion on the PR will be lost to the sands fo time. Please put a comment here!

nairbv replied:
@ezyang is there a better way though? as long as we're converting a double to an integral type there needs to be a cast, and at least now we've properly range checked it.

ezyang replied:
I couldn't think of a better way to fix it, which is why I did not ask you to fix it :)

vadimkantorov replied:
FWIW this seems to be the boost way: https://www.boost.org/doc/libs/1_57_0/libs/numeric/conversion/doc/html/boost_numericconversion/improved_numeric_cast__.html (maybe its source code has a relatively self-contained function)

ezyang reviewed (approved):
okey dokey

nairbv committed:
remove unnecessary cast

nairbv committed:
Revert "remove unnecessary cast"

This reverts commit eba160c316f4b03f3c4653b33b802412c1ad8596.
